### PR TITLE
Enabled migration support

### PR DESCRIPTION
**Pull request by cutig3r**
@smitpatel, here is the pull request to add migration support.  ~~I have no clue how to actually get it to test the generated code, so I'm sure there are more tests needed. The one that I did add was adapted from [EFCore.Pluralizer](https://github.com/bricelam/EFCore.Pluralizer/blob/041b2a13a3076499f0fd51b9dc8a7eb4d266b127/EFCore.Pluralizer.Test/EFCorePluralizerServicesTests.cs#L10)~~

This is a partial replacement for #12 and addresses #9

### CODE REVIEW

File: EFCore.SqlServer.HierarchyId/ChangeTracking/SqlServerHierarchyIdValueComparer.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    ///         will require both if mutation is to be detected.
+    ///     </para>
+    ///     <para>
+    ///         Snapshotting is the process of creating a copy of the value into a snapshot so it can
+    ///         later be compared to determine if it has changed. For some types, such as collections,
+    ///         this needs to be a deep copy of the collection rather than just a shallow copy of the
+    ///         reference.
+    ///     </para>
+    /// </summary>
+    public class SqlServerHierarchyIdValueComparer : ValueComparer<HierarchyId>
```
Comment (smitpatel) [inline EFCore.SqlServer.HierarchyId/ChangeTracking/SqlServerHierarchyIdValueComparer.cs:21]: Why a custom value comparer is required for migration support?

**Comment by cutig3r** [comment]
I've added additional tests for snapshots and migrations, adapted from [ModelSnapshotSqlServerTest.cs](https://github.com/dotnet/efcore/blob/v3.1.0/test/EFCore.Design.Tests/Migrations/ModelSnapshotSqlServerTest.cs) and [CSharpMigrationOperationGeneratorTest.cs](https://github.com/dotnet/efcore/blob/v3.1.0/test/EFCore.Design.Tests/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs).  The tests are skipped on .NET Framework because the underlying test code doesn't support it.

File: EFCore.SqlServer.HierarchyId/Storage/SqlServerHierarchyIdTypeMapping.cs [hunk 2]
```diff
@@ -2,48 +2,63 @@
...
     internal class SqlServerHierarchyIdTypeMapping : RelationalTypeMapping
     {
         private static readonly MethodInfo _getSqlBytes
-            = typeof(SqlDataReader).GetTypeInfo().GetDeclaredMethod(nameof(SqlDataReader.GetSqlBytes));
+            = typeof(SqlDataReader).GetRuntimeMethod(nameof(SqlDataReader.GetSqlBytes), new[] { typeof(int) });
 
         private static Action<DbParameter, SqlDbType> _sqlDbTypeSetter;
         private static Action<DbParameter, string> _udtTypeNameSetter;
 
+        protected virtual ValueConverter<HierarchyId, SqlBytes> ValueConverter { get; }
```
[the remark this answers is not in the document]
Comment (cutig3r) [reply EFCore.SqlServer.HierarchyId/Storage/SqlServerHierarchyIdTypeMapping.cs]: @smitpatel, I narrowed it down to why it isn't being called when you pass in the converter.  It comes down to [this line in CHarpHelper.cs](https://github.com/dotnet/efcore/blob/82c6ea483d3a17393dad5986df9a8eef89ddcd07/src/EFCore.Design/Design/Internal/CSharpHelper.cs#L805).  This ends up calling FindMapping without providing a StoreTypeName and at this point, the literalType is SqlBytes.  In [SqlServerHierarchyIdTypeMappingSourcePlugin.cs](https://github.com/efcore/EFCore.SqlServer.HierarchyId/blob/a77a339308d09893820edbac9b2d6e21850980dd/EFCore.SqlServer.HierarchyId/Storage/SqlServerHierarchyIdTypeMappingSourcePlugin.cs#L14), it only wants to provide a mapping if the type is HierarchyId or if the StoreTypeName is "hierarchyid", and the mapping info with a ClrType of SqlBytes and a null StoreTypeName will not satisfy this condition  
By not providing a converter, SqlBytes will never be passed to the FindMapping method.

[the remark this answers is not in the document]
Comment (smitpatel) [reply EFCore.SqlServer.HierarchyId/Storage/SqlServerHierarchyIdTypeMapping.cs]: Thanks for the update @cutig3r. Currently I am busy in some work in efcore repo. I will come back to this soon. Looks like something needs to be changed to identify the type mapping correctly either here in plugin or in EF Core CSharpHelper.

File: EFCore.SqlServer.HierarchyId.Test/Migrations/NeedsBuildReferencesFixture.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public class NeedsBuildReferencesFixture
+    {
+        public NeedsBuildReferencesFixture()
+        {
+            AssemblyUtils.SetEntryAssembly();
+        }
+
+        private IEnumerable<BuildReference> _getRequiredReferences()
+        {
+            if (AssemblyUtils.WasEntryAssemblyUpdated)
```
Comment (bricelam) [inline EFCore.SqlServer.HierarchyId.Test/Migrations/NeedsBuildReferencesFixture.cs:20]: This seems like overkill. Do you really need to update the entry assembly? Can't we just use `#if NET472`?

File: EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
+﻿// Adapted From: https://github.com/dotnet/efcore/blob/v3.1.0/test/EFCore.Design.Tests/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Comment (bricelam) [inline EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs:3]: Hmm, not sure we're allowed to relicense that under the MIT...

File: EFCore.SqlServer.HierarchyId.Test/Migrations/NeedsBuildReferencesFixture.cs (hunk 3 above)
Comment (cutig3r) [reply EFCore.SqlServer.HierarchyId.Test/Migrations/NeedsBuildReferencesFixture.cs:20]: I modified the entry assembly to get the DependencyContext.Default to not be null (referenced in BuildReference.cs), without having to make changes to the code from dotnet/efcore (everything in the TestUtilities folder).  I've tried experimenting without it, and ran into referencing issues, but can dive deeper if necessary.

File: EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs (hunk 4 above)
Comment (cutig3r) [reply EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs:3]: Yeah, that is a big question.  Everything in the TestUtilities folder was copied from dotnet/efcore (in an effort to keep it fairly easy to merge to dotnet/efcore).  Should these migration tests be removed?

Comment (bricelam) [reply EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs:3]: Merging this back into EF Core seems like it'll be an even bigger nightmare. 😅 Let's cross that bridge when we get there.

Personally, I think most of this testing infrastructure is probably unnecessary for this repo. Maybe just simplify it to what's absolutely needed and call it an original work. 🤷‍♂️

Comment (cutig3r) [reply EFCore.SqlServer.HierarchyId.Test/Migrations/Design/CSharpMigrationOperationGeneratorTest.cs]: Migration tests have now been simplified

**Comment by cutig3r** [comment]
Let me know if there's anything else that needs to change.  Thanks!

**Comment by cutig3r** [comment]
@smitpatel or @bricelam 
Is this ok to merge?

**Pull request merged**